DiscordBot.Database/DatabaseContainer.cs

```diff
@@ -1,60 +1,83 @@
-﻿using System.Linq;
+﻿using System;
+using System.Linq;
 using System.Threading.Tasks;
-using DiscordBot.Domain.Configuration;
 using Microsoft.Azure.Cosmos;
-using Microsoft.Extensions.Options;
 
 namespace DiscordBot.Database
 {
     public class DatabaseContainer<T> where T : DatabaseObject
     {
+        private readonly ObjectStoreProperties _props;
         private readonly string _name;
         private readonly Microsoft.Azure.Cosmos.Database _database;
-        private readonly ObjectStoreProperties _props;
         private Container _container;
-        public Container Container;
 
-        public DatabaseContainer(IOptions<DatabaseSettings> configuration, string name, CosmosClient client, ObjectStoreProperties props)
+        public DatabaseContainer(IDatabaseService databaseService)
         {
-            if (name != null && name != "")
-            {
-                _name = name;
-            }
+            _props = DatabaseHelpers.GetProperties<T>();
+            _name = _props.ContainerName;
+
+            if (databaseService.Client == null)
+                _database = null;
             else
-            {
-                _name = props.ContainerName;
-            }
-            _database = client.GetDatabase(configuration.Value.Name);
-            _props = props;
-            InitAsync().Wait();
+                _database = databaseService.Client.GetDatabase(databaseService.Configuration.Value.Name);
+
+            _container = null;
         }
         private async Task InitAsync()
```

Comment:
You should return ValueTask. This speeds things up.

Reply:
Should I do that everywhere?

Reply:
No. The reason why we use ValueTask is, that a majority of the times, this method gets executed, it won't actually do async code.
ValueTask is a ValueType aka struct. 

Structs are faster when they stay on the Stack.
"How do values stay on the Stack?" you may ask. - It's simple. Variables that live inside a function, are on the stack. Variables/fields inside a class are on the heap.


This method will be slower than normal Task for the first execution, because thats when it transitions to the async part, the ValueTask will be put on the Heap.